File: Solution-Python/Binary_94.py
class Solution:
    # Using Stack
    def inorderTraversal(self, root: TreeNode) -> List[int]:
        res, stack, node = [], [], root
        while stack or node:
            if node:
                stack.append(node)
                node = node.left
            else:
                node = stack.pop()
                res.append(node.val)
                node = node.right
        return res



    # Morris Traversal
    def inorderTraversal(self, root: TreeNode) -> List[int]:
        res, node = [], root
        while node:
            if node.left:
                pred = node.left
                while pred.right and pred.right != node:
                    pred = pred.right
                if pred.right:
                    pred.right = None
                    res.append(node.val)
                    node = node.right
                else:
                    pred.right = node
                    node = node.left
            else:
                res.append(node.val)
                node = node.right
        return res

File: Solution-Go/Binary_94.go
// Using Stack
func inorderTraversal(root *TreeNode) []int {
    res, stack, node := []int{}, []*TreeNode{}, root
    for len(stack) > 0 || node != nil {
        if node != nil {
            stack = append(stack, node)
            node = node.Left
        } else {
            node = stack[len(stack) - 1]
            stack = stack[:len(stack) - 1]
            res = append(res, node.Val)
            node = node.Right
        }
    }
    return res
}



// Morris Traversal
func inorderTraversal(root *TreeNode) []int {
    res, node := []int{}, root
    for node != nil {
        if node.Left == nil {
            res = append(res, node.Val)
            node = node.Right
        } else {
            pred := node.Left
            for pred.Right != nil && pred.Right != node {
                pred = pred.Right
            }
            if pred.Right == nil {
                pred.Right = node
                node = node.Left
            } else {
                pred.Right = nil
                res = append(res, node.Val)
                node = node.Right
            }
        }
    }
    return res
}